### PR TITLE
App codebase improvement

### PR DESCRIPTION
**Pull request by BobbyESP**
- Moved hard-coded strings to string resources
- Moved UI states, Query states and LyricsStatus to enum classes
- Added kotlinx.serialization
- Added savers for serialization for both Song and SongInfo data classes
- Added saver for Uri's
- Fixed lyrics being cleared after rotating the screen (Note for the dev: using a LaunchedEffect you make sure that the rememberSaveable values are accessible)

### CODE REVIEW

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 1]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                }, confirmButton = {
+                    Button(onClick = {
+                        Thread {
+                            queryStatus = QueryStatus.LyricsPending
+                            try {
+                                lyricsResult =
+                                    viewModel.getSyncedLyrics(queryResult.songLink.toString())
+                                queryStatus = QueryStatus.LyricsSuccess
+                            } catch (e: Exception) {
+                                if (e is FileNotFoundException) lyricsResult = "No lyrics found"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:324]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 2]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
-                        ) {
-                            Text(text = "Try again")
-                        }
+                AlertDialog(title = {
+                    Text(text = stringResource(R.string.song_info_fetched))
+                }, text = {
+                    Column {
+                        Text(
+                            text = stringResource(
+                                R.string.song_name, songNameResult ?: "Unknown"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:303]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 3]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                            )
+                        )
+                        Text(
+                            text = stringResource(
+                                R.string.artist_name, artistNameResult ?: "Unknown"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:308]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 4]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
-                )
+
+            QueryStatus.Pending -> {
+                AlertDialog(title = {
+                    Text(text = stringResource(R.string.fetching_song_info))
+                }, text = {
+                    Text(
+                        text = stringResource(
+                            R.string.fetching_song_info_for_by,
+                            song.title ?: "Unknown",
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:283]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 5]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                            song.artist ?: "Unknown"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:284]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 6]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
-                                    else {
+
+            QueryStatus.NotSubmitted -> {
+                AlertDialog(title = {
+                    Text(text = stringResource(R.string.get_lyrics))
+                }, text = {
+                    Text(
+                        text = stringResource(
+                            R.string.selected_song_by,
+                            song.title ?: "Unknown",
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:245]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 7]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                            song.artist ?: "Unknown"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:246]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 8]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
                     },
-                    title = { Text(text = "Error") },
+                    title = { Text(text = stringResource(id = R.string.error)) },
                     text = {
-                        if (failReason.contains("NotFound") || failReason.contains("JSON"))
-                            Text(text = "No results")
-                        else
-                            Text(text = "An error occurred: $failReason")
+                        if (failReason.contains("NotFound") || failReason.contains("JSON")) Text(
+                            text = "No results"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:430]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 9]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                    Button(onClick = {
+                        val lrc =
+                            "[ti:${queryResult.songName}]\n" + "[ar:${queryResult.artistName}]\n" + "[by:Generated using SongSync]\n" + lyricsResult
+                        val file = File(
+                            song.filePath?.dropLast(4) + ".lrc"
+                        )
+                        file.writeText(lrc)
+
+                        Toast.makeText(
+                            context, "Lyrics saved to ${file.path}", Toast.LENGTH_LONG
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:390]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 10]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
-                                Toast.LENGTH_LONG
-                            ).show()
+            QueryStatus.LyricsPending -> {
+                AlertDialog(title = {
+                    Text(text = stringResource(R.string.fetching_lyrics))
+                }, text = {
+                    Text(
+                        text = stringResource(
+                            R.string.fetching_lyrics_for_by,
+                            queryResult.songName ?: "Unknown",
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:358]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 11]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                            queryResult.artistName ?: "Unknown"
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:359]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 12]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+        UiState.Pending -> {
+            AlertDialog(title = {
+                Text(text = stringResource(id = R.string.batch_download_lyrics))
+            }, text = {
+                Column {
+                    Text(text = stringResource(R.string.downloading_lyrics))
+                    MarqueeText(
+                        stringResource(
+                            R.string.song,
+                            songs[(successCount + failedCount) % total].title ?: "Unknown",
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:500]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 13]
```diff
@@ -235,323 +231,310 @@ fun SongItem(song: Song, viewModel: MainViewModel) {
...
+                    Text(text = stringResource(R.string.please_do_not_close_the_app_this_may_take_a_while))
                 }
-            )
+            }, onDismissRequest = {
+                uiState = UiState.Cancelled
+            }, confirmButton = {
+                // no button but compose cries when i don't use confirmButton
+            }, dismissButton = {
+                OutlinedButton(onClick = { uiState = UiState.Cancelled }) {
+                    Text(text = "Cancel")
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:524]: hardcoded

File: app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt [hunk 14]
```diff
@@ -585,65 +568,63 @@ fun BatchDownloadLyrics(songs: List<Song>, viewModel: MainViewModel, onDone: ()
...
+                    Text(text = stringResource(R.string.download_complete))
+                    Text(text = stringResource(R.string.success, successCount))
+                    Text(text = stringResource(R.string.failed, failedCount))
                 }
-            )
+            }, onDismissRequest = {
+                uiState = UiState.Cancelled
+            }, confirmButton = {
+                Button(onClick = { uiState = UiState.Cancelled }) {
+                    Text(text = "OK")
```
Comment (nift4) [inline app/src/main/java/pl/lambada/songsync/ui/screens/HomeScreen.kt:595]: hardcoded

**Review by nift4 — COMMENTED** [review]
some strings were missed

File: app/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -42,4 +42,24 @@
     <string name="save_lrc_file">Save .lrc file</string>
     <string name="no_results">No results</string>
     <string name="no_internet_server">No internet connection or server is down</string>
+    <string name="lyrics_not_found">Lyrics not found</string>
+    <string name="unknown">Unknown</string>
+    <string name="lyrics_saved_to">Lyrics saved to</string>
+    <string name="generated_using">Generated using SongSync</string>
```
Comment (nift4) [inline app/src/main/res/values/strings.xml:48]: mark translatable="false", this should not be localized (may cause issues with utf8 langs like persian, chinese, etc), is inside file so should not be dependant on UI language

Comment (BobbyESP) [reply app/src/main/res/values/strings.xml:48]: Okay!

**Comment by Lambada10** [comment]
App crashes when trying to batch download lyrics

**Comment by BobbyESP** [comment]
> App crashes when trying to batch download lyrics

Yeah, same to me, I saw it before, but seems to be a problem while trying to parse some characters. Can you tell me if it occurs in your version?

**Comment by BobbyESP** [comment]
I'm not at home right now, so I send this to you

**Comment by BobbyESP** [comment]
Seems to be related with some string in the values.xml file. Later I fix it, sorry

**Comment by BobbyESP** [comment]
> App crashes when trying to batch download lyrics

Fixed

**Pull request merged**